src/components/withPickerEvents.js

```diff
@@ -0,0 +1,120 @@
+import React, {Component} from 'react';
+import {pickerPropTypes, pickerDefaultProps} from './Picker/pickerPropTypes';
+
+export default function (WrappedComponent) {
+    class WithSharedEvents extends Component {
```

Comment:
I haven't delved deeply into the reasons for needing it, but is an HOC necessary? Feels like the "WrappedComponent" is always going to be the `Picker`.

Additionally, I'm having trouble following some of the conventions used here e.g. the `callbackConsumer()` (stuck out to me as unusual for our codebase and my instincts are telling me it could confuse people).

I'd have to do a deeper dive to come up with more concrete suggestions, but just wanted to toss that observation out there that maybe we could simplify this.

Reply:
The HOC wraps the Picker in both entrypoints (index.js and index.native.js), so that we can share all of the code contained in it between them. This also gives us a way to manipulate the events passed down to BasePicker using  an entrypoint file. I originally had placed the HOC in the Picker directory because it would only ever be used for the Picker, but I noticed all of the others were contained in the base `src` directory so I moved it there.

An HOC would seem like overkill but I am kind of stuck between a dead simple solution (the one I originally provided), and an over-engineered solution.  I have been having trouble finding something in the middle ground that won't introduce a lot of code duplication.  Another option might be class inheritance.

We basically need a way to share all of the event handlers between both entrypoints but give the web entry point a way to extend/override them before passing them down to BasePicker.  If you have any alternative ideas for that please let me know. If we stick with the HOC I can definitely rename `callbackConsumer` back to `onInputChange`.

Reply:
Ok I see the issue now - sorry for the distraction here. I think where this is at is a bit hard to reason about for not much gain.

Here's what I'm thinking as an alternative...

1. Start by reverting what you have here back to https://github.com/Expensify/App/pull/15673/commits/15fb9c7db99ac31f24b6b78926115a1a4cca5e8b
1. Create some prop like `shouldUseAdditionalPickerEvents` on the Picker
1. Override the methods that way inside the Picker

It might seem vain, but we can get around the cross platform file extension convention this way and it will allow web/native to switch on/off the functionality we need without specifically referencing the platform anywhere.

Reply:
@marcaaron Thank you for the advice.  I just need a bit more clarification:

> It might seem vain, but we can get around the cross platform file extension convention this way and it will allow web/native to switch on/off the functionality we need without specifically referencing the platform anywhere.

Are you suggesting not having platform-specific entry points in this case?  I am just curious what is going to toggle `shouldUseAdditionalPickerEvents` if we don't, and if we didn't, where do we define the web-specific events?

Reply:
@marcaaron Ok I will give that a shot.  Since it's just a conditional in the entry points, is this implying that we keep the web events defined in the Picker itself?

Reply:
>implying that we keep the web events defined in the Picker itself

That was my meaning yea.

But I thought of something that is slightly better and might result in fewer changes while avoiding a reference to the web specific `e.target`...

- Create `getAdditionalPickerEvents()`
    - have native return `{}`
    - have web return something like

```js
// getAdditionalPickerEvents/index.js
function getAdditionalPickerEvents(onMouseDown, onChange) {
    return {
        onMouseDown,
        onChange: (e) = {
            if (e.target.selectedIndex === undefined) {
                return;
            }
            const index = e.target.selectedIndex;
            const value = e.target.options[index].value;
            onChange(value, index);
        },
    };
}
```

Then pass them like:
```js
pickerProps={{
    ...getAdditionalPickerEvents(this.setIsOpen, this.updateAndClose),
```
  
This feels slightly better to me as it manages the web event outside of the cross platform Picker code. Sorry for all the back and forth. Our style guide sometimes paints us into a corner 😄

Reply:
@marcaaron I like the looks of that.  I have been trying to find a way to pass down the events from the web entry, but also to be able to pass a function back that can enable/disable the highlight without having to reference them from inside the shared Picker.  This looks like it could work.

Also, no worries. I am happy to work towards something we can all be satisfied with.